src/analyses/variable-sensitivity/abstract_pointer_object.cpp

```diff
@@ -87,6 +100,16 @@ abstract_object_pointert abstract_pointer_objectt::write_dereference(
   return std::make_shared<abstract_pointer_objectt>(type(), true, false);
 }
 
+abstract_object_pointert abstract_pointer_objectt::typecast(
+  const typet &new_type,
+  const abstract_environmentt &environment,
+  const namespacet &ns) const
+{
+  INVARIANT(is_void_pointer(type()), "Only allow pointer casting from void*");
```

Comment:
I feel kind of conflicted about this.  In some ways it should not be an `INVARIANT` because it can be triggered by a user but also, at the moment we really aren't handling this case.  Could we return a `TOP` pointer of the correct type?  Reads from that should give `TOP` and writes to it could call `havoc()`?

Reply:
I am similarly conflicted. Right now, the call to typecast() is itself guarded so that it will never be called if this invariant will fire

```
  if(expr.id() == ID_typecast)
  {
    const typecast_exprt &tce = to_typecast_expr(expr);
    if(tce.op().id() == ID_symbol && is_void_pointer(tce.op().type()))
``` 

Consequently existing typecasts, of which there are a handful, will evaluate as they do at the moment. I suppose this invariant is more of warning for future care than a guard against wacky corner cases.